exporter/metric/prometheus/prometheus.go

```diff
@@ -0,0 +1,384 @@
+	if err != nil {
+		return nil, err
+	}
+
+	e.counters[mKey] = counter
+	return counter, nil
+}
+
+func (e *Exporter) getGauge(record export.Record, mKey metricKey) (prometheus.Gauge, error) {
+	e.Lock()
+	defer e.Unlock()
+	if g, ok := e.gauges[mKey]; ok {
+		return g, nil
+	}
+
+	desc := record.Descriptor()
+	gaugeVec, err := e.getGaugeVec(desc, record.Labels())
+	if err != nil {
+		return nil, err
+	}
+
+	gauge, err := gaugeVec.GetMetricWithLabelValues(labelsToTags(record.Labels())...)
+	if err != nil {
+		return nil, err
+	}
+	e.gauges[mKey] = gauge
+	return gauge, nil
+}
+
+func (e *Exporter) getHistogram(record export.Record, mKey metricKey) (prometheus.Observer, error) {
+	e.Lock()
+	defer e.Unlock()
+	if g, ok := e.histograms[mKey]; ok {
+		return g, nil
+	}
+
+	desc := record.Descriptor()
+	histogramVec, err := e.getHistogramVec(desc, record.Labels())
+	if err != nil {
+		return nil, err
+	}
+
+	histogram, err := histogramVec.GetMetricWithLabelValues(labelsToTags(record.Labels())...)
+	if err != nil {
+		return nil, err
+	}
+	e.histograms[mKey] = histogram
+	return histogram, nil
+}
+
+func (e *Exporter) getCounterVec(desc *export.Descriptor, labels export.Labels) (*prometheus.CounterVec, error) {
+	if c, ok := e.counterVecs[desc]; ok {
+		return c, nil
+	}
+
+	tagKeys := getTagKeys(labels.Ordered())
+	c := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: Sanitize(desc.Name()),
+			Help: desc.Description(),
+		},
+		tagKeys,
+	)
+
+	if err := e.registerer.Register(c); err != nil {
+		return nil, err
+	}
+
+	e.counterVecs[desc] = c
+	return c, nil
+}
+
+func (e *Exporter) getGaugeVec(desc *export.Descriptor, labels export.Labels) (*prometheus.GaugeVec, error) {
+	if g, ok := e.gaugeVecs[desc]; ok {
+		return g, nil
+	}
+
+	tagKeys := getTagKeys(labels.Ordered())
+	g := prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: Sanitize(desc.Name()),
+			Help: desc.Description(),
+		},
+		tagKeys,
+	)
+
+	if err := e.registerer.Register(g); err != nil {
+		return nil, err
+	}
+
+	e.gaugeVecs[desc] = g
+	return g, nil
+}
+
+func (e *Exporter) getHistogramVec(desc *export.Descriptor, labels export.Labels) (*prometheus.HistogramVec, error) {
+	if g, ok := e.histogramVecs[desc]; ok {
+		return g, nil
+	}
+
+	tagKeys := getTagKeys(labels.Ordered())
+	g := prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    Sanitize(desc.Name()),
+			Help:    desc.Description(),
+			Buckets: e.defaultHistogramBuckets,
+		},
+		tagKeys,
+	)
+
+	if err := e.registerer.Register(g); err != nil {
+		return nil, err
+	}
+
+	e.histogramVecs[desc] = g
+	return g, nil
+}
+
+func getTagKeys(keys []core.KeyValue) []string {
+	tagKeys := make([]string, 0, len(keys))
+	for _, kv := range keys {
+		tagKeys = append(tagKeys, Sanitize(string(kv.Key)))
+	}
+	return tagKeys
+}
+
+func labelsToTags(labels export.Labels) []string {
+	tags := make([]string, 0, labels.Len())
+	for _, label := range labels.Ordered() {
```

Comment:
I could imagine avoiding the calls to `Emit()` below by parsing sub-strings out of `labels.Encoded()`. Maybe that's crazy.

On a related note, I just did a thought exercise about what it would take so that this `[]string` can be precomputed in the batcher.  The defaultkeys batcher has computed a `[]core.KeyValue` with the correct size and order. Computing a separate slice of strings at the place where this is done could potentially be a win for this exporter. (E.g., it could be an optional field in the `export.Record`, and instead of having LabelEncoder return a string, it could return both a string and a []string). Nevermind!

Reply:
I guess we could parse the encoded label. Are you okay leaving this to another PR?

Reply:
Definitely, or a TODO that's never done. :)